### PR TITLE
Add handling for sympy conditions in deferred measurement transformer

### PR DESCRIPTION
**Pull request by daxfohl**
Previously the DeferredMeasurementTransformer could only work with classical conditions of type KeyCondition (triggered iff the measurement value of that key != zero). This PR enables the transformer to work with SympyConditions (which allow arbitrary classical logic with multiple keys in the trigger) as well.

We do this by calculating the truth table of the sympy condition, and constructing a quantum control that has the same truth table for the deferred measurement circuit. This approach actually applies to not only sympy conditions, but all classical control conditions. Thus we removed the existing logic for KeyCondition transformation.

Additionally this adds a control_values reduction in the ControlledGate constructor allowing us to remove the special case logic from the transformer.

### CODE REVIEW

**Review by tanujkhattar — COMMENTED** [review]
@daxfohl Can you please add a description describing the change in this PR ?

File: cirq-core/cirq/transformers/measurement_transformers.py [hunk 1]
```diff
@@ -139,6 +147,37 @@ def defer(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
     return circuit
 
 
+def _all_possible_datastore_states(
+    keys: Iterable['cirq.MeasurementKey'],
```
Comment (daxfohl) [inline cirq-core/cirq/transformers/measurement_transformers.py:151]: Needs to be a set. It's not well-defined if there's a dupe. (i.e. condition is `m==m`.

**Comment by daxfohl** [comment]
Converting to draft until #5883 is approved

File: cirq-core/cirq/transformers/measurement_transformers.py [hunk 2]
```diff
@@ -109,25 +119,34 @@ def defer(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
...
+                if key not in measurement_qubits:
+                    raise ValueError(f'Deferred measurement for key={key} not found.')
+
+            # Try every possible datastore state against the condition, and the ones that work
+            # are the control values for the new op.
+            compatible_datastores = [
+                store
+                for store in _all_possible_datastore_states(keys, measurement_qubits)
+                if all(c.resolve(store) for c in op.classical_controls)
+            ]
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/measurement_transformers.py:139]: I think it's worth highlighting the exponential complexity in the docstring above.

**Review by tanujkhattar — APPROVED** [review]
LGTM % nits

**Pull request merged**